pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -34,87 +34,127 @@ import (
+	"k8s.io/kubernetes/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/util/wait"
 	"k8s.io/kubernetes/pkg/util/workqueue"
 	"k8s.io/kubernetes/pkg/watch"
 )
 
 const ResourceResyncTime = 30 * time.Second
 
-type Monitor struct {
-	Store      cache.Store
-	Controller *framework.Controller
+type monitor struct {
+	store      cache.Store
+	controller *framework.Controller
 }
 
-type SelfReference struct {
+type objectReference struct {
 	metatypes.OwnerReference
 	// This is needed by the dynamic client
 	Namespace string
 }
 
-func (s SelfReference) String() string {
-	return fmt.Sprintf("%s/%s, namespace: %s, name: %s, uid: %s", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
+func (s objectReference) String() string {
+	return fmt.Sprintf("[%s/%s, namespace: %s, name: %s, uid: %s]", s.APIVersion, s.Kind, s.Namespace, s.Name, s.UID)
 }
 
+// node does not require a lock to protect. The single-threaded
+// Propagator.processEvent() is the sole writer of the nodes. The multi-threaded
+// GarbageCollector.processItem() reads the nodes, but it only reads the fields
+// that never gets changed by Propagator.processEvent().
 type node struct {
-	identity   SelfReference
-	dependents []*node
+	identity   objectReference
+	dependents map[*node]struct{}
 	// When processing an Update event, we need to compare the updated
 	// ownerReferences with the owners recorded in the graph.
 	owners []metatypes.OwnerReference
 }
 
-type EventType string
+type eventType string
 
-const Add EventType = "Add"
-const Update EventType = "Update"
-const Delete EventType = "Delete"
+const addEvent eventType = "Add"
+const updateEvent eventType = "Update"
+const deleteEvent eventType = "Delete"
 
-type Event struct {
-	Type   EventType
+type event struct {
+	Type   eventType
 	Obj    interface{}
 	OldObj interface{}
 }
 
 type Propagator struct {
-	eventQueue  *workqueue.Type
-	graphLookup map[types.UID]*node
-	gc          *GarbageCollector
+	eventQueue *workqueue.Type
+	// uidToNode doesn't require a lock, because only the single-threaded Propagator.processEvent() read/write it.
+	uidToNode map[types.UID]*node
+	gc        *GarbageCollector
 }
 
-func (p *Propagator) addToGraphLookup(n *node) {
+func (p *Propagator) insertNode(n *node) {
 	fmt.Println("CHAO: add to graph:", n.identity)
-	p.graphLookup[n.identity.UID] = n
+	p.uidToNode[n.identity.UID] = n
 }
 
-func (p *Propagator) removeFromGraphLookup(n *node) {
+func (p *Propagator) removeNode(n *node) {
 	fmt.Println("CHAO: remove from graph:", n.identity)
-	delete(p.graphLookup, n.identity.UID)
+	delete(p.uidToNode, n.identity.UID)
 }
 
-func (p *Propagator) updateOwners(n *node) (ownerExists bool) {
-	fmt.Println("CHAO: in updateOwners, graphLookUp:", p.graphLookup)
+// addToOwners finds all owners as listed in n.owners, and marks n as one of
+// their dependents. ownerExists indicates if any owner listed in n.owners exist
+// in p.uidToNode.
+func (p *Propagator) addToOwners(n *node) (ownerExists bool) {
```

Comment:
Is it a good idea to have addToOwners and insertNode be separate steps? (same for remove)